akka-stream/src/main/scala/akka/stream/impl/streamref/SourceRefImpl.scala

```diff
@@ -172,103 +193,312 @@ private[stream] final class SourceRefStageImpl[Out](val initialPartnerRef: Optio
-
-          self.unwatch(sender)
-          completed = true
-          tryPush()
+          state match {
+            case Running(partner) =>
+              // upstream completed, continue running until we have emitted every element in buffer
+              // or downstream cancels
+              verifyPartner(sender, partner)
+              log.debug(
+                "[{}] The remote stream has completed, emitting {} elements left in buffer before completing",
+                stageActorName,
+                receiveBuffer.used)
+              self.unwatch(sender)
+              state = UpstreamCompleted(partner)
+              tryPush()
+            case WaitingForCancelAck(_, _) =>
+              // upstream completed while we were waiting for it to receive cancellation and ack
+              // upstream may stop without seeing cancellation, but we may not see termination
+              // let the cancel timeout hit
+              log.debug("[{}] Upstream completed while waiting for cancel ack", stageActorName)
+            case other =>
+              // UpstreamCompleted, AwaitingPartner or AwaitingSubscription(_) all means a bug here
+              throw new IllegalStateException(
+                s"[$stageActorName] Saw RemoteStreamCompleted($seqNr) while in state $other, should never happen")
+          }
 
         case (sender, StreamRefsProtocol.RemoteStreamFailure(reason)) =>
-          observeAndValidateSender(sender, "Illegal sender in RemoteSinkFailure")
-          log.warning("[{}] The remote stream has failed, failing (reason: {})", stageActorName, reason)
+          state match {
+            case weKnoPartner: WeKnowPartner =>
+              val partner = weKnoPartner.partner
+              verifyPartner(sender, partner)
+              log.debug("[{}] The remote stream has failed, failing (reason: {})", stageActorName, reason)
+              failStage(
+                RemoteStreamRefActorTerminatedException(
+                  s"[$stageActorName] Remote stream (${sender.path}) failed, reason: $reason"))
+            case other =>
+              throw new IllegalStateException(
+                s"[$stageActorName] got RemoteStreamFailure($reason) when in state $other, should never happen")
+          }
 
-          self.unwatch(sender)
-          failStage(RemoteStreamRefActorTerminatedException(s"Remote stream (${sender.path}) failed, reason: $reason"))
+        case (sender, StreamRefsProtocol.Ack) =>
+          state match {
+            case WaitingForCancelAck(partner, cause) =>
+              verifyPartner(sender, partner)
+              log.debug(s"[$stageActorName] Got cancellation ack from remote, canceling", stageActorName)
+              cancelStage(cause)
+            case other =>
+              throw new IllegalStateException(s"[$stageActorName] Got an Ack when in state $other")
+          }
 
         case (_, Terminated(p)) =>
```

Comment:
you know about the somewhat "surprising" situation when the Terminated may arrive before the last message is sent from the same actor? taking different lanes in Artery. I know we have some "workaround" for that in stream refs, and just wanted to mention it so that this refactoring is taking that into consideration.

Reply:
Yep, that's part of the trixery